packages/babel-types/scripts/generators/validators.ts

```diff
@@ -48,11 +48,20 @@ function addIsHelper(
   }
 
   const result =
+    NODE_FIELDS[type] || FLIPPED_ALIAS_KEYS[type]
+      ? `node is t.${type}`
+      : "boolean";
+
+  const resultWithOpts =
     NODE_FIELDS[type] || FLIPPED_ALIAS_KEYS[type]
       ? `node is t.${type} & Opts`
       : "boolean";
 
-  return `export function is${type}<Opts extends Options<t.${type}>>(node: t.Node | null | undefined, opts?: Opts | null): ${result} {
+  return [
+    // Signature overload to avoid issues like https://github.com/babel/babel/pull/17503#discussion_r2325598609
+    `export function is${type}(node: t.Node | null | undefined): ${result};`,
+    `export function is${type}<Opts extends Options<t.${type}>>(node: t.Node | null | undefined, opts?: Opts | null): ${resultWithOpts};`,
+    `export function is${type}<Opts extends Options<t.${type}>>(node: t.Node | null | undefined, opts?: Opts | null): ${resultWithOpts} {
```

Comment:
This last one can probably just return `boolean` now, since its type is not actually exposed externally but only used to check the return type.